libcst/metadata/reentrant_codegen.py

```diff
@@ -0,0 +1,210 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+# pyre-strict
+
+from dataclasses import dataclass, field
+from typing import List, Optional, Sequence
+
+from libcst import BaseStatement, CSTNode, Module
+from libcst._add_slots import add_slots
+from libcst._nodes.internal import CodegenState
+from libcst.metadata import BaseMetadataProvider
+
+
+class CodegenPartial:
+    """
+    Provided by :class:`ExperimentalReentrantCodegenProvider`.
+
+    Stores enough information to generate either a small patch
+    (:meth:`get_modified_code_range`) or a new file (:meth:`get_modified_code`) by
+    replacing the old node at this position.
+    """
+
+    __slots__ = [
+        "start_offset",
+        "end_offset",
+        "has_trailing_newline",
+        "_indent_tokens",
+        "_prev_codegen_state",
+    ]
+
+    def __init__(self, state: "_ReentrantCodegenState") -> None:
+        # store a frozen copy of these values, since they change over time
+        self.start_offset: int = state.start_offset_stack[-1]
+        self.end_offset: int = state.char_offset
+        self.has_trailing_newline: bool = True  # this may get updated to False later
+        self._indent_tokens: Sequence[str] = tuple(state.indent_tokens)
+        # everything else can be accessed from the codegen state object
+        self._prev_codegen_state: _ReentrantCodegenState = state
+
+    def get_original_module_code(self) -> str:
+        """
+        Equivalent to :meth:`libcst.Module.bytes` on the top-level module that contains
+        this statement, except that it uses the cached result from our previous code
+        generation pass, so it's faster.
+        """
+        return self._prev_codegen_state.get_code()
+
+    def get_original_module_bytes(self) -> bytes:
+        """
+        Equivalent to :meth:`libcst.Module.bytes` on the top-level module that contains
+        this statement, except that it uses the cached result from our previous code
+        generation pass, so it's faster.
+        """
+        return self.get_original_module_code().encode(self._prev_codegen_state.encoding)
+
+    def get_original_node_code(self) -> str:
+        """
+        Equivalent to :meth:`libcst.Module.code_for_node` on the current statement,
+        except that it uses the cached result from our previous code generation pass,
+        so it's faster.
+        """
+        return self._prev_codegen_state.get_code()[self.start_offset : self.end_offset]
+
+    def get_modified_node_code(self, node: BaseStatement) -> str:
```

Comment:
Should this be named `get_modified_statement_code` since it doesn't work on any node type?